### PR TITLE
@types/rosie: allow factory.build with class constructor

### PR DESCRIPTION
**Pull request by chrislgrigg**
The existing definition wasn't quite right. I'm not sure how to write the definition to make this type safe, but this will at least let it compile until we can get a better fix. Having trouble running tests locally but I think this should work. I'll remove the error that's thrown, if not.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/rosiejs/rosie#associate-a-factory-with-an-existing-class>>

### CODE REVIEW

File: types/rosie/index.d.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ declare namespace rosie {
      * @param {function(object): *=} constructor
      * @return {Factory}
      */
-    define<T = any>(name: string, constructor?: (opts?: any) => any): IFactory<T>;
+    define<T = any>(name: string, constructor?: any): IFactory<T>;
```
Comment (andy-ms) [inline types/rosie/index.d.ts]: Could the type be `((opts?: any) => any) | (new(opts?: any) => any))`?

File: types/rosie/rosie-tests.ts [hunk 2]
```diff
@@ -65,6 +65,18 @@ personFactory.attr('secretCode', ['firstName', 'lastName', 'age', 'age', 'firstN
...
+  type: string;
+
+  constructor(props: { type: string }) {
+    this.type = props.type;
+  }
+}
+
+const customTestFactory = Factory.define<CustomClass>('test', CustomClass);
+const output = customTestFactory.build({ type: 'foo' });
+if (output.type !== 'foo') { throw new Error('constructor did not receive args'); }
```
Comment (andy-ms) [inline types/rosie/rosie-tests.ts]: The test code doesn't actually run, so this seems unnecessary? It compiled before this change.

File: types/rosie/index.d.ts (hunk 1 above)
Comment (subvertallchris) [reply types/rosie/index.d.ts]: Hey Andy! I was unclear on how to express it and what was there wasn't working, this is an attempt to at least let it compile and preserve functionality. Is that the syntax for a an object that has a constructor?

Comment (andy-ms) [reply types/rosie/index.d.ts]: Yep, `new(opts?: any) => any` is equivalent to `{ new(opts?: any): any }`.

Comment (subvertallchris) [reply types/rosie/index.d.ts]: Ah, that's really cool. I'll update this when I get a chance, probably tomorrow. Thank you!

File: types/rosie/rosie-tests.ts (hunk 2 above)
Comment (subvertallchris) [reply types/rosie/rosie-tests.ts]: Oh, is that so? I didn't realize that. It makes sense, now that you point it out. :-) I'll remove that when I make the other change.

**Comment by andy-ms** [comment]
Thanks!

**Pull request merged**